File: Sources/Entity/BedrockEntityCommonNBT.swift
import Foundation

/// Common Bedrock entity tags used by the empty-entity creator and the summon
/// command. Type-specific tags are intentionally left to copied/imported NBT.
enum BedrockEntityCommonNBT {
    static func tags(
        identifier: String,
        position: BedrockWorldObjectPosition,
        dimension: Int32,
        uniqueID: Int64
    ) -> [NBTNamedTag] {
        [
            NBTNamedTag(name: "definitions", value: .list(.string, [.string("+\(identifier)")])),
            NBTNamedTag(name: "UniqueID", value: .long(uniqueID)),
            NBTNamedTag(name: "Air", value: .short(300)),
            NBTNamedTag(name: "Chested", value: .byte(0)),
            NBTNamedTag(name: "Color", value: .byte(0)),
            NBTNamedTag(name: "Color2", value: .byte(0)),
            NBTNamedTag(name: "CustomName", value: .string("")),
            NBTNamedTag(name: "CustomNameVisible", value: .byte(0)),
            NBTNamedTag(name: "Dead", value: .byte(0)),
            NBTNamedTag(name: "DeathTime", value: .short(0)),
            NBTNamedTag(name: "DimensionId", value: .int(dimension)),
            NBTNamedTag(name: "FallDistance", value: .float(0)),
            NBTNamedTag(name: "Fire", value: .short(0)),
            NBTNamedTag(name: "HasBoundOrigin", value: .byte(0)),
            NBTNamedTag(name: "Invulnerable", value: .byte(0)),
            NBTNamedTag(name: "IsAngry", value: .byte(0)),
            NBTNamedTag(name: "IsAutonomous", value: .byte(0)),
            NBTNamedTag(name: "IsBaby", value: .byte(0)),
            NBTNamedTag(name: "IsEating", value: .byte(0)),
            NBTNamedTag(name: "IsGliding", value: .byte(0)),
            NBTNamedTag(name: "IsGlobal", value: .byte(0)),
            NBTNamedTag(name: "IsIllagerCaptain", value: .byte(0)),
            NBTNamedTag(name: "IsOrphaned", value: .byte(0)),
            NBTNamedTag(name: "IsOutOfControl", value: .byte(0)),
            NBTNamedTag(name: "IsRoaring", value: .byte(0)),
            NBTNamedTag(name: "IsScared", value: .byte(0)),
            NBTNamedTag(name: "IsStunned", value: .byte(0)),
            NBTNamedTag(name: "IsSwimming", value: .byte(0)),
            NBTNamedTag(name: "IsTamed", value: .byte(0)),
            NBTNamedTag(name: "IsTrusting", value: .byte(0)),
            NBTNamedTag(name: "LastDimensionId", value: .int(dimension)),
            NBTNamedTag(name: "LootDropped", value: .byte(0)),
            NBTNamedTag(name: "MarkVariant", value: .int(0)),
            NBTNamedTag(name: "Motion", value: .list(.float, [.float(0), .float(0), .float(0)])),
            NBTNamedTag(name: "OnGround", value: .byte(1)),
            NBTNamedTag(name: "OwnerNew", value: .long(-1)),
            NBTNamedTag(name: "Persistent", value: .byte(1)),
            NBTNamedTag(name: "PortalCooldown", value: .int(0)),
            NBTNamedTag(name: "Pos", value: .list(.float, [
                .float(Float(position.x)), .float(Float(position.y)), .float(Float(position.z))
            ])),
            NBTNamedTag(name: "Rotation", value: .list(.float, [.float(0), .float(0)])),
            NBTNamedTag(name: "Saddled", value: .byte(0)),
            NBTNamedTag(name: "Sheared", value: .byte(0)),
            NBTNamedTag(name: "ShowBottom", value: .byte(0)),
            NBTNamedTag(name: "Sitting", value: .byte(0)),
            NBTNamedTag(name: "SkinID", value: .int(0)),
            NBTNamedTag(name: "Strength", value: .int(0)),
            NBTNamedTag(name: "StrengthMax", value: .int(0)),
            NBTNamedTag(name: "Tags", value: .list(.string, [])),
            NBTNamedTag(name: "Variant", value: .int(0))
        ]
    }

    static func addingMissingTopLevel(_ defaults: [NBTNamedTag], to root: NBTValue) throws -> NBTValue {
        guard case .compound(var tags) = root else {
            throw MCBEEditorError.malformedData("实体 NBT 根必须是 Compound")
        }
        var existing = Set(tags.map { $0.name.lowercased() })
        for tag in defaults where existing.insert(tag.name.lowercased()).inserted {
            tags.append(tag)
        }
        return .compound(tags)
    }

    static func mergingTopLevel(_ additions: [NBTNamedTag], into root: NBTValue) throws -> NBTValue {
        guard case .compound(var tags) = root else {
            throw MCBEEditorError.malformedData("实体 NBT 根必须是 Compound")
        }
        for addition in additions {
            if let index = tags.firstIndex(where: { $0.name.caseInsensitiveCompare(addition.name) == .orderedSame }) {
                tags[index] = addition
            } else {
                tags.append(addition)
            }
        }
        return .compound(tags)
    }

    static func identifier(in root: NBTValue) -> String? {
        if let raw = root.stringValue(namedAny: ["identifier", "Identifier", "id", "Id"]), raw.contains(":") {
            return normalizedIdentifier(raw)
        }
        if let definitions = root.value(namedAny: ["definitions", "Definitions"])?.listValues {
            for value in definitions.reversed() {
                guard case .string(var text) = value else { continue }
                while text.first == "+" || text.first == "-" { text.removeFirst() }
                if text.contains(":") { return normalizedIdentifier(text) }
            }
        }
        if let numeric = root.int64Value(namedAny: ["id", "Id"]),
           let entry = BedrockDataValueCatalog.entity(forNumericID: numeric) {
            return entry.identifier
        }
        return nil
    }

    static func position(in root: NBTValue) -> BedrockWorldObjectPosition? {
        guard let values = root.value(namedAny: ["Pos", "pos", "Position", "position"])?.listValues,
              values.count >= 3,
              let x = values[0].numericDoubleValue,
              let y = values[1].numericDoubleValue,
              let z = values[2].numericDoubleValue else { return nil }
        return BedrockWorldObjectPosition(x: x, y: y, z: z)
    }

    static func dimension(in root: NBTValue) -> Int32? {
        root.int64Value(namedAny: ["DimensionId", "DimensionID", "Dimension", "dimension"])
            .map { Int32(clamping: $0) }
    }

    static func uniqueID(in root: NBTValue) -> Int64? {
        root.int64Value(namedAny: ["UniqueID", "UniqueId", "uniqueID", "uniqueId"])
    }

    private static func normalizedIdentifier(_ value: String) -> String {
        let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        return trimmed.contains(":") ? trimmed : "minecraft:\(trimmed)"
    }
}

File: Windows/MCBEEditor.Core/Entity/BedrockEntityCommonNbt.cs
using MCBEEditor.Core.Nbt;

namespace MCBEEditor.Core.Entity;

/// <summary>
/// Common Bedrock entity tags shared by the Windows entity creator/importer.
/// Type-specific data is intentionally preserved from copied/imported NBT.
/// </summary>
public static class BedrockEntityCommonNbt
{
    public static IReadOnlyList<NbtNamedTag> Tags(string identifier, BedrockWorldObjectPosition position, int dimension, long uniqueId)
        =>
        [
            new("definitions", new NbtListValue(NbtTagType.String, [new NbtStringValue("+" + NormalizeIdentifier(identifier))])),
            new("UniqueID", new NbtLongValue(uniqueId)),
            new("Air", new NbtShortValue(300)),
            new("Chested", new NbtByteValue(0)),
            new("Color", new NbtByteValue(0)),
            new("Color2", new NbtByteValue(0)),
            new("CustomName", new NbtStringValue(string.Empty)),
            new("CustomNameVisible", new NbtByteValue(0)),
            new("Dead", new NbtByteValue(0)),
            new("DeathTime", new NbtShortValue(0)),
            new("DimensionId", new NbtIntValue(dimension)),
            new("FallDistance", new NbtFloatValue(0)),
            new("Fire", new NbtShortValue(0)),
            new("HasBoundOrigin", new NbtByteValue(0)),
            new("Invulnerable", new NbtByteValue(0)),
            new("IsAngry", new NbtByteValue(0)),
            new("IsAutonomous", new NbtByteValue(0)),
            new("IsBaby", new NbtByteValue(0)),
            new("IsEating", new NbtByteValue(0)),
            new("IsGliding", new NbtByteValue(0)),
            new("IsGlobal", new NbtByteValue(0)),
            new("IsIllagerCaptain", new NbtByteValue(0)),
            new("IsOrphaned", new NbtByteValue(0)),
            new("IsOutOfControl", new NbtByteValue(0)),
            new("IsRoaring", new NbtByteValue(0)),
            new("IsScared", new NbtByteValue(0)),
            new("IsStunned", new NbtByteValue(0)),
            new("IsSwimming", new NbtByteValue(0)),
            new("IsTamed", new NbtByteValue(0)),
            new("IsTrusting", new NbtByteValue(0)),
            new("LastDimensionId", new NbtIntValue(dimension)),
            new("LootDropped", new NbtByteValue(0)),
            new("MarkVariant", new NbtIntValue(0)),
            new("Motion", new NbtListValue(NbtTagType.Float, [new NbtFloatValue(0), new NbtFloatValue(0), new NbtFloatValue(0)])),
            new("OnGround", new NbtByteValue(1)),
            new("OwnerNew", new NbtLongValue(-1)),
            new("Persistent", new NbtByteValue(1)),
            new("PortalCooldown", new NbtIntValue(0)),
            new("Pos", PositionTag(position)),
            new("Rotation", new NbtListValue(NbtTagType.Float, [new NbtFloatValue(0), new NbtFloatValue(0)])),
            new("Saddled", new NbtByteValue(0)),
            new("Sheared", new NbtByteValue(0)),
            new("ShowBottom", new NbtByteValue(0)),
            new("Sitting", new NbtByteValue(0)),
            new("SkinID", new NbtIntValue(0)),
            new("Strength", new NbtIntValue(0)),
            new("StrengthMax", new NbtIntValue(0)),
            new("Tags", new NbtListValue(NbtTagType.String, [])),
            new("Variant", new NbtIntValue(0))
        ];

    public static NbtListValue PositionTag(BedrockWorldObjectPosition position)
        => new(NbtTagType.Float,
        [
            new NbtFloatValue(CheckedFloat(position.X, "X")),
            new NbtFloatValue(CheckedFloat(position.Y, "Y")),
            new NbtFloatValue(CheckedFloat(position.Z, "Z"))
        ]);

    public static NbtValue AddMissingTopLevel(NbtValue root, IEnumerable<NbtNamedTag> defaults)
    {
        if (root is not NbtCompoundValue compound) throw new InvalidDataException("实体 NBT 根必须是 Compound。");
        var tags = compound.Tags.Select(tag => new NbtNamedTag(tag.Name, NbtDocumentTools.DeepClone(tag.Value))).ToList();
        var names = new HashSet<string>(tags.Select(tag => tag.Name), StringComparer.OrdinalIgnoreCase);
        foreach (var tag in defaults)
            if (names.Add(tag.Name)) tags.Add(new NbtNamedTag(tag.Name, NbtDocumentTools.DeepClone(tag.Value)));
        return new NbtCompoundValue(tags);
    }

    public static string? Identifier(NbtValue root)
    {
        var direct = root.CompoundValueIgnoreCase("identifier", "Identifier", "id", "Id");
        if (direct is NbtStringValue text && !string.IsNullOrWhiteSpace(text.Value))
        {
            var raw = text.Value.Trim();
            if (raw.Contains(':')) return NormalizeIdentifier(raw);
            if (long.TryParse(raw, out var numericText)) return BedrockEntityCatalog.IdentifierForNumericId(numericText);
        }
        if (direct?.IntegerValue() is long numeric) return BedrockEntityCatalog.IdentifierForNumericId(numeric);
        if (root.CompoundValueIgnoreCase("definitions", "Definitions") is NbtListValue list)
        {
            for (var i = list.Values.Count - 1; i >= 0; i--)
            {
                if (list.Values[i] is not NbtStringValue definition) continue;
                var value = definition.Value.TrimStart('+', '-').Trim();
                if (value.Contains(':')) return NormalizeIdentifier(value);
            }
        }
        return null;
    }

    public static BedrockWorldObjectPosition? Position(NbtValue root)
        => BedrockWorldObjectScanner.ExtractPosition(root, BedrockWorldObjectKind.Entity);

    public static int? Dimension(NbtValue root) => BedrockWorldObjectScanner.ExtractDimension(root);

    public static long? UniqueId(NbtValue root)
        => root.CompoundValueIgnoreCase("UniqueID", "UniqueId", "uniqueID", "uniqueId")?.IntegerValue();

    public static string NormalizeIdentifier(string identifier)
    {
        var value = identifier.Trim().ToLowerInvariant();
        return value.Contains(':') ? value : "minecraft:" + value;
    }

    private static float CheckedFloat(double value, string axis)
    {
        if (!double.IsFinite(value) || value < -float.MaxValue || value > float.MaxValue)
            throw new InvalidDataException($"实体 {axis} 坐标无法写入 Float。");
        return (float)value;
    }
}
